atp/client.go

```diff
@@ -370,7 +371,79 @@ func (c *client) sendErrorToAll(err error) {
 	c.mutex.Unlock()
 }
 
-//nolint:funlen
+func (c *client) handleWorkDoneMessage(runtimeMessage DecodedRuntimeMessage) {
+	var doneMessage WorkDoneMessage
+	if err := cbor.Unmarshal(runtimeMessage.RawMessageData, &doneMessage); err != nil {
+		c.logger.Errorf("Failed to decode work done message (%v) for run ID '%s' ", err, runtimeMessage.RunID)
+		c.mutex.Lock()
+		c.sendExecutionResult(runtimeMessage.RunID, NewErrorExecutionResult(
+			fmt.Errorf("failed to decode work done message (%w)", err)))
+		c.mutex.Unlock()
+		return
+	}
+	c.mutex.Lock()
+	c.sendExecutionResult(runtimeMessage.RunID, c.processWorkDone(runtimeMessage.RunID, doneMessage))
+	c.mutex.Unlock()
+}
+
+func (c *client) handleSignalMessage(runtimeMessage DecodedRuntimeMessage) {
+	var signalMessage SignalMessage
+	if err := cbor.Unmarshal(runtimeMessage.RawMessageData, &signalMessage); err != nil {
+		c.logger.Errorf("ATP client for run ID '%s' failed to decode signal message: %v",
+			runtimeMessage.RunID, err)
+	}
+	c.mutex.Lock()
+	signalChannel, found := c.runningStepEmittedSignalChannels[runtimeMessage.RunID]
+	c.mutex.Unlock()
+	if !found {
+		c.logger.Warningf(
+			"Step with run ID '%s' sent signal '%s'. Ignoring; signal handling is not implemented "+
+				"(emittedSignals is nil).",
+			runtimeMessage.RunID, signalMessage.SignalID)
+	} else {
+		c.logger.Debugf("Got signal from step with run ID '%s' with ID '%s'", runtimeMessage.RunID,
+			signalMessage.SignalID)
+		signalChannel <- signalMessage.ToInput(runtimeMessage.RunID)
+	}
+}
+
+// Returns true if fatal, requiring aborting the read loop.
+func (c *client) handleErrorMessage(runtimeMessage DecodedRuntimeMessage) bool {
+	var errMessage ErrorMessage
+	if err := cbor.Unmarshal(runtimeMessage.RawMessageData, &errMessage); err != nil {
+		c.logger.Errorf("Step with run ID '%s' failed to decode error message: %v",
+			runtimeMessage.RunID, err)
+	}
+	c.logger.Errorf("Step with run ID '%s' sent error message: %v", runtimeMessage.RunID, errMessage)
+	resultMsg := fmt.Errorf("step '%s' sent error message: %s", runtimeMessage.RunID,
+		errMessage.ToString(runtimeMessage.RunID))
+	if errMessage.ServerFatal {
+		c.sendErrorToAll(resultMsg)
+		return true // It's server fatal, so this is the last message from the server.
+	} else if errMessage.StepFatal {
+		if runtimeMessage.RunID == "" {
+			c.sendErrorToAll(fmt.Errorf("step fatal error missing run id (%w)", resultMsg))
+		} else {
+			c.sendExecutionResult(runtimeMessage.RunID, NewErrorExecutionResult(resultMsg))
+		}
+	}
+	return false
```

Comment:
If it's not fatal, it's only logged. This is client (engine) side, so it will end up in the correct logs.

I think this format with the items indented under StepFatal makes sense. I don't see any benefit from that refactor.

The RunID in this case only matters for step fatal messages, so I don't think any change is required there.